File: box/src/main/java/de/qabel/box/storage/command/UpdateFileChange.java
package de.qabel.box.storage.command;

import de.qabel.box.storage.BoxFile;
import de.qabel.box.storage.BoxFolder;
import de.qabel.box.storage.BoxObject;
import de.qabel.box.storage.DirectoryMetadata;
import de.qabel.box.storage.exceptions.QblStorageException;
import de.qabel.box.storage.exceptions.QblStorageNameConflict;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class UpdateFileChange implements DirectoryMetadataChange<UpdateFileChange.Result> {
    private static final Logger logger = LoggerFactory.getLogger(UpdateFileChange.class);
    private BoxFile oldFile;
    private BoxFile newFile;

    public UpdateFileChange(BoxFile oldFile, BoxFile newFile) {
        this.oldFile = oldFile;
        this.newFile = newFile;
    }

    @Override
    public Result execute(DirectoryMetadata dm) throws QblStorageException {
        String filename = newFile.getName();
        try {
            dm.insertFile(newFile);
        } catch (QblStorageNameConflict e) {
                BoxObject currentFile = findCurrentFileOrFolder(dm, filename);
                deleteObject(currentFile, dm);
                dm.insertFile(newFile);
                while (true) {
                    try {
                        logger.debug("Conflicting " + filename);
                        filename = filename + "_conflict";
                        logger.debug("Inserting conflict marked file as " + filename);
                        currentFile.setName(filename);
                        insertObject(currentFile, dm);
                        break;
                    } catch (QblStorageNameConflict ignored) {
                    }
                }
        }
        return new Result();
    }

    private BoxObject findCurrentFileOrFolder(DirectoryMetadata dm, String filename) throws QblStorageException {
        BoxObject currentFile = dm.getFile(filename);
        if (currentFile == null) {
            currentFile = dm.getFolder(filename);
        }
        return currentFile;
    }

    private void insertObject(BoxObject currentFile, DirectoryMetadata dm) throws QblStorageException {
        if (currentFile instanceof BoxFile) {
            dm.insertFile((BoxFile) currentFile);
        } else if (currentFile instanceof BoxFolder) {
            dm.insertFolder((BoxFolder) currentFile);
        }
    }

    private void deleteObject(BoxObject currentObject, DirectoryMetadata dm) throws QblStorageException {
        if (currentObject instanceof BoxFile) {
            dm.deleteFile((BoxFile) currentObject);
        } else if (currentObject instanceof BoxFolder) {
            dm.deleteFolder((BoxFolder) currentObject);
        }
    }

    public static class Result {

    }
}

File: box/src/main/java/de/qabel/box/storage/command/UpdateFileChange.kt
package de.qabel.box.storage.command

import de.qabel.box.storage.*
import de.qabel.box.storage.exceptions.QblStorageNameConflict
import org.slf4j.LoggerFactory

open class UpdateFileChange(
    val expectedFile: BoxFile?,
    val newFile: BoxFile
    ) : DMChange<Unit>, Postprocessable {

    private val logger by lazy { LoggerFactory.getLogger(UpdateFileChange::class.java) }

    private var sameFileByHash = false

    private fun hasSameHash(obj: BoxObject) = obj is BoxFile && obj.isHashed() && obj.hashed == newFile.hashed

    override fun postprocess(dm: DirectoryMetadata, writeBackend: StorageWriteBackend, shares: ShareHolder) {
        if (sameFileByHash) {
            writeBackend.deleteBlock(newFile.block)
        }
    }

    override fun execute(dm: DirectoryMetadata) {
        var filename = newFile.name
        try {
            dm.getFile(newFile.name)?.apply {
                if (isSame(expectedFile)) {
                    dm.deleteFile(this)
                }
            }
            dm.insertFile(newFile)
        } catch (e: QblStorageNameConflict) {
            with (findCurrentFileOrFolder(dm, filename)) {
                if (hasSameHash(this)) {
                    sameFileByHash = true
                    return
                }
                deleteObject(this, dm)
                dm.insertFile(newFile)
                while (true) {
                    try {
                        logger.debug("Conflicting " + filename)
                        filename += "_conflict"
                        logger.debug("Inserting conflict marked file as " + filename)
                        this.name = filename
                        insertObject(this, dm)
                        break
                    } catch (ignored: QblStorageNameConflict) {
                    }
                }
            }
        }
    }

    private fun findCurrentFileOrFolder(dm: DirectoryMetadata, filename: String)
        = dm.getFile(filename)
            ?: dm.getFolder(filename)
            ?: throw IllegalStateException("conflicting name is neither file nor folder: " + filename);

    private fun insertObject(currentObject: BoxObject, dm: DirectoryMetadata)
        = when (currentObject) {
            is BoxFile -> dm.insertFile(currentObject)
            is BoxFolder -> dm.insertFolder(currentObject)
            else -> throw NotImplementedError("not implemented for " + currentObject.javaClass)
        }

    private fun deleteObject(currentObject: BoxObject, dm: DirectoryMetadata)
        = when (currentObject) {
            is BoxFile -> dm.deleteFile(currentObject)
            is BoxFolder -> dm.deleteFolder(currentObject)
            else -> throw NotImplementedError("not implemented for " + currentObject.javaClass)
        }
}
